crates/chain/src/local_chain.rs

```diff
@@ -17,224 +152,339 @@ impl ChainOracle for LocalChain {
                 _ => None,
             },
         )
     }
-}
 
-impl AsRef<BTreeMap<u32, BlockHash>> for LocalChain {
-    fn as_ref(&self) -> &BTreeMap<u32, BlockHash> {
-        &self.blocks
+    fn get_chain_tip(&self) -> Result<Option<BlockId>, Self::Error> {
+        Ok(self.checkpoints.values().last().map(CheckPoint::block_id))
     }
 }
 
-impl From<LocalChain> for BTreeMap<u32, BlockHash> {
-    fn from(value: LocalChain) -> Self {
-        value.blocks
+impl LocalChain {
+    /// Construct a [`LocalChain`] from an initial `changeset`.
+    pub fn from_changeset(changeset: ChangeSet) -> Self {
+        let mut chain = Self::default();
+        chain.apply_changeset(&changeset);
+        chain
     }
-}
 
-impl From<BTreeMap<u32, BlockHash>> for LocalChain {
-    fn from(value: BTreeMap<u32, BlockHash>) -> Self {
-        Self { blocks: value }
+    /// Construct a [`LocalChain`] from a given `checkpoint` tip.
+    pub fn from_checkpoint(checkpoint: CheckPoint) -> Self {
+        Self {
+            checkpoints: checkpoint.iter().map(|cp| (cp.height(), cp)).collect(),
+        }
     }
-}
 
-impl LocalChain {
-    pub fn from_blocks<B>(blocks: B) -> Self
-    where
-        B: IntoIterator<Item = BlockId>,
-    {
+    /// Constructs a [`LocalChain`] from a [`BTreeMap`] of height to [`BlockHash`].
+    ///
+    /// The [`BTreeMap`] enforces the height order. However, the caller must ensure the blocks are
+    /// all of the same chain.
+    pub fn from_blocks(blocks: BTreeMap<u32, BlockHash>) -> Self {
         Self {
-            blocks: blocks.into_iter().map(|b| (b.height, b.hash)).collect(),
+            checkpoints: blocks
+                .into_iter()
+                .map({
+                    let mut prev = None;
+                    move |(height, hash)| {
+                        let cp = CheckPoint::new_with_prev(BlockId { height, hash }, prev.clone())
+                            .expect("must not fail");
+                        prev = Some(cp.clone());
+                        (height, cp)
+                    }
+                })
+                .collect(),
         }
     }
 
-    /// Get a reference to a map of block height to hash.
-    pub fn blocks(&self) -> &BTreeMap<u32, BlockHash> {
-        &self.blocks
+    /// Get the highest checkpoint.
+    pub fn tip(&self) -> Option<CheckPoint> {
+        self.checkpoints.values().last().cloned()
     }
 
-    pub fn tip(&self) -> Option<BlockId> {
-        self.blocks
-            .iter()
-            .last()
-            .map(|(&height, &hash)| BlockId { height, hash })
+    /// Returns whether the [`LocalChain`] is empty (has no checkpoints).
+    pub fn is_empty(&self) -> bool {
+        self.checkpoints.is_empty()
     }
 
-    /// This is like the sparsechain's logic, expect we must guarantee that all invalidated heights
-    /// are to be re-filled.
-    pub fn determine_changeset(&self, update: &Self) -> Result<ChangeSet, UpdateNotConnectedError> {
```

Comment:
Don't you still need this? You want to `determine_changeset`, apply to persistent storage and then update in memory if it succeeds. In order to only have `update` we'd have to make `LocalChain` monotone. We could do this after this PR pretty easily I think and so leave it as `update` if we promise to fix it later.

Reply:
`update` returns a closure first. The local chain only updates when the returned closure is called. It is done in this way so we can make use of the pointers during the update.

Reply:
I don't really like this. I can just call `apply_changeset` instead of calling this closure to do the same thing?. It'd be fine with just applying the changeset right away and trying to fix #1005 eventually.

Reply:
Yes, but you won't get some of the benefits of the linked list. What do you not like about this API?

Reply:
It's redundant which makes it confusing. Is the point I might decide not to apply it? In which situation would I not want to apply it?

Reply:
Ah I see what you mean, I'll remove the closure and have the method apply changes right away.

Reply:
Done.